src/statistics.rs

```diff
@@ -0,0 +1,771 @@
+struct MessageStats {
+    to_verified: u32,
+    unverified_encrypted: u32,
+    unencrypted: u32,
+    only_to_self: u32,
+}
+
+#[repr(u32)]
+#[derive(Debug, Clone, Copy, FromPrimitive, FromSql, PartialEq, Eq, PartialOrd, Ord)]
+enum SecurejoinSource {
+    /// Because of some problem, it is unknown where the QR code came from.
+    Unknown = 0,
+    /// The user opened a link somewhere outside Delta Chat
+    ExternalLink = 1,
+    /// The user clicked on a link in a message inside Delta Chat
+    InternalLink = 2,
+    /// The user clicked "Paste from Clipboard" in the QR scan activity
+    Clipboard = 3,
+    /// The user clicked "Load QR code as image" in the QR scan activity
+    ImageLoaded = 4,
+    /// The user scanned a QR code
+    Scan = 5,
+}
+
+#[derive(Serialize)]
+struct SecurejoinSources {
+    unknown: u32,
+    external_link: u32,
+    internal_link: u32,
+    clipboard: u32,
+    image_loaded: u32,
+    scan: u32,
+}
+
+#[derive(Debug, Clone, Copy, FromPrimitive, FromSql, PartialEq, Eq, PartialOrd, Ord)]
+enum SecurejoinUIPath {
+    /// The UI path is unknown, or the user didn't open the QR code screen at all.
+    Unknown = 0,
+    /// The user directly clicked on the QR icon in the main screen
+    QrIcon = 1,
+    /// The user first clicked on the `+` button in the main screen,
+    /// and then on "New Contact"
+    NewContact = 2,
+}
+
+#[derive(Serialize)]
+struct SecurejoinUIPaths {
+    other: u32,
+    qr_icon: u32,
+    new_contact: u32,
+}
+
+/// Some information on an invite-joining event
+/// (i.e. a qr scan or a clicked link).
+#[derive(Serialize)]
+struct JoinedInvite {
+    /// Whether the contact already existed before.
+    /// If this is false, then a contact was newly created.
+    already_existed: bool,
+    /// If a contact already existed,
+    /// this tells us whether the contact was verified already.
+    already_verified: bool,
+    /// The type of the invite:
+    /// "contact" for 1:1 invites that setup a verified contact,
+    /// "group" for invites that invite to a group
+    /// and also perform the contact verification 'along the way'.
+    typ: String,
+}
+
+/// Sends a message with statistics about the usage of Delta Chat,
+/// if the last time such a message was sent
+/// was more than a week ago.
+///
+/// On the other end, a bot will receive the message and make it available
+/// to Delta Chat's developers.
+pub async fn maybe_send_statistics(context: &Context) -> Result<Option<ChatId>> {
+    if should_send_statistics(context).await? {
+        let last_sending_time = context.get_config_i64(Config::StatsLastSent).await?;
+        let next_sending_time = last_sending_time.saturating_add(SENDING_INTERVAL_SECONDS);
+        if next_sending_time <= time() {
+            // If something goes wrong, try again in 1 minute.
+            // This prevents infinite loops in the (unlikely) case of an error:
+            let one_minute_later = last_sending_time.saturating_add(60).to_string();
```

Comment:
I think that even 12 hours would be fine, it's still much less than a week. Retrying every minute if smth is completely broken looks excessive

Reply:
Another problem here is that if `send_statistics()` succeeds, but then the program crashes, next time we'll send statistics for messages only for this minute. We indeed shouldn't send differential statistics, it's even hard to think about all possible edge cases, considering that there are also backups/restorations.